cmd/node-cache/app/cache_app.go

```diff
@@ -0,0 +1,242 @@
+package app
+
+import (
+	"fmt"
+	"os"
+
+	"k8s.io/dns/cmd/kube-dns/app/options"
+	"k8s.io/dns/pkg/dns/config"
+
+	"net"
+	"strings"
+	"time"
+
+	"github.com/coredns/coredns/coremain"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"k8s.io/dns/pkg/netif"
+	"k8s.io/kubernetes/pkg/util/dbus"
+	utilexec "k8s.io/kubernetes/pkg/util/exec"
+	utiliptables "k8s.io/kubernetes/pkg/util/iptables"
+)
+
+// ConfigParams lists the configuration options that can be provided to node-cache
+type ConfigParams struct {
+	LocalIPStr           string        // comma separated listen ips for the local cache agent
+	LocalIPs             []net.IP      // parsed ip addresses for the local cache agent to listen for dns requests
+	LocalPort            string        // port to listen for dns requests
+	MetricsListenAddress string        // address to serve metrics on
+	InterfaceName        string        // Name of the interface to be created
+	Interval             time.Duration // specifies how often to run iptables rules check
+	BaseCoreFile         string        // Path to the template config file for node-cache
+	CoreFile             string        // Path to config file used by node-cache
+	KubednsCMPath        string        // Directory where kube-dns configmap will be mounted
+	UpstreamSvcName      string        // Name of the service whose clusterIP is the upstream for node-cache for cluster domain
+	SetupIptables        bool
+}
+
+type iptablesRule struct {
+	table utiliptables.Table
+	chain utiliptables.Chain
+	args  []string
+}
+
+// CacheApp contains all the config required to run node-cache.
+type CacheApp struct {
+	iptables      utiliptables.Interface
+	iptablesRules []iptablesRule
+	params        *ConfigParams
+	netifHandle   *netif.NetifManager
+	kubednsConfig *options.KubeDNSConfig
+	exitChan      chan bool // Channel to terminate background goroutines
+	clusterDNSIP  net.IP
+}
+
+func isLockedErr(err error) bool {
+	return strings.Contains(err.Error(), "holding the xtables lock")
+}
+
+// Init initializes the parameters and networking setup necessary to run node-cache
+func (c *CacheApp) Init() {
+	c.netifHandle = netif.NewNetifManager(c.params.LocalIPs)
+	if c.params.SetupIptables {
+		c.initIptables()
+	}
+	err := c.TeardownNetworking()
+	if err != nil {
+		// It is likely to hit errors here if previous shutdown cleaned up all iptables rules and interface.
+		// Logging error at info level
+		clog.Infof("Hit error during teardown - %s", err)
+	}
+	err = c.setupNetworking()
+	if err != nil {
+		c.TeardownNetworking()
+		clog.Fatalf("Failed to setup - %s, Exiting", err)
+	}
+	initMetrics(c.params.MetricsListenAddress)
+	c.updateCorefile(&config.Config{})
```

Comment:
Curious why do we need to update it once with an empty config? It seems `initKubeDNSConfigSync` also updates it with the initialConfig.

Reply:
Ah I see, that may not run if configmap path is not given.

Reply:
exactly. added a comment here to clarify this.